pyhap/accessory_driver.py

```diff
@@ -859,118 +862,98 @@ def set_characteristics(self, chars_query, client_addr):
         :type chars_query: dict
         """
         # TODO: Add support for chars that do no support notifications.
-        updates = {}
-        setter_results = {}
-        setter_responses = {}
-        had_error = False
-        had_write_response = False
-        expired = False
 
+        queries: List[Dict[str, Any]] = chars_query[HAP_REPR_CHARS]
+
+        self._notify(queries, client_addr)
+
+        updates_by_accessories_services: Dict[
+            Accessory, Dict[Service, Dict[Characteristic, Any]]
+        ] = defaultdict(lambda: defaultdict(dict))
+        results: Dict[int, Dict[int, dict[str, Any]]] = defaultdict(
```

Comment:
```suggestion
        results: Dict[int, Dict[int, Dict[str, Any]]] = defaultdict(
```

Reply:
one more (I always forget these since I'm used to working with py3.11 which has so much better typing!)